lib/cache.ml

```diff
@@ -0,0 +1,59 @@
+open Bos
+open Rresult
+open Astring
+
+let pp_header = Fmt.(styled `Blue string)
+
+let header = "==> "
+
+let valid_remote_name = String.map (function '/' | '@' | ':' -> '-' | x -> x)
+
+let git_remote_exists_or_create ~remotes ~remote =
+  let remote_name = valid_remote_name remote in
+  match String.Set.mem remote_name remotes with
+  | true -> Ok ()
+  | false -> Exec.git_remote_add ~remote ~remote_name
+
+let git_branch_exists_or_create ~output_dir ~remotes ~branch ~remote ~tag cache_dir =
+  let remote_name = valid_remote_name remote in
+  OS.Dir.with_current cache_dir
+    (fun () ->
+      OS.Path.exists Fpath.(v ".git" / "refs" / "heads" / branch) >>= function
+      | true ->
+          Logs.app (fun l ->
+              l "%a%aPulling sources for %a." pp_header header
+                Fmt.(styled `Green string)
+                "[CACHED] "
+                Fmt.(styled `Cyan Fpath.pp)
+                output_dir );
+          Ok ()
+      | false -> (
+          Logs.app (fun l ->
+              l "%aPulling sources for %a." pp_header header Fmt.(styled `Cyan Fpath.pp) output_dir
+          );
+          git_remote_exists_or_create ~remotes ~remote >>= function
+          | () -> Exec.git_fetch_to ~remote_name ~tag ~branch ) )
+    ()
+  |> Rresult.R.join
+
+let git_get ~output_dir ~remote ~tag () =
+  let home = Unix.getenv "HOME" in
+  let cache_location = Fpath.(v home / ".duniverse" / "cache") in
```

Comment:
Probably worth plumbing through `DUNIVERSE_CACHE` env var to let the user set it easily.  In CI, we can map it to a volume mount in a docker container then to be shared easily across CI runs.

Reply:
Also it's worth using XDG conventions to avoid cluttering HOME with dot files/dirs. For instance using: 
```
$XDG_CACHE_HOME defines the base directory relative to which user specific non-essential data files should be stored. If $XDG_CACHE_HOME is either not set or empty, a default equal to $HOME/.cache should be used.
```
(from https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)

Reply:
FYI there's some code in dune that implements this.

Reply:
So now it checks for a `DUNIVERSE_CACHE`, then for a `XDG_CACHE_HOME` and finally uses `HOME` to find a cache location !